### PR TITLE
Added Subscription emitter

### PR DESCRIPTION
**Pull request by AvivBenShoham**
## Proposed changes (including videos or screenshots)

Added Lock according to user to prevent bugs during 2 critical section while user is being logged in/logged out.
Added interval to prevent the cost of unsub then immediate sub on client refreshes.

Added Support for subscription using redis pub sub

### CODE REVIEW

File: app/redis/handleRedisMessage.ts [hunk 1]
```diff
@@ -1,11 +1,25 @@
...
+
+export const redisMessageHandlers: Partial<IRedisHandlers> = {};
 
 redis.on('message', (channel: string, msg: string) => {
-  console.log('new message from redis');
-  
-  const message = superjson.parse(msg);
-  const ns = message.ns; 
-  return redisMessageHandlers[ns](message);
+	console.log('new message from redis');
```
Comment (LeeorShalev) [inline app/redis/handleRedisMessage.ts:16]: remove log

File: app/ws/connectionMappings.ts [hunk 2]
```diff
@@ -25,37 +25,80 @@ const updateMappingsOnSub = (connectionId: string, channels: Set<string>, userId
...
 };
 
-const decreaseChannelsBinding = (connectionId: string): void => {
+const decreaseConnectionIdFromChannelBindCount = async (channel: string, connectionId: string): Promise<void> => {
+	const release = await acquireLock(channel);
+	try {
+		const listeningConnections = channelListeners.get(channel);
+		console.log('listeners ', listeningConnections, channel, connectionId);
+		if (listeningConnections?.size === 1) {
+			console.log(`Unsubscribing to channel: ${ channel }`);
```
Comment (LeeorShalev) [inline app/ws/connectionMappings.ts:42]: remove log

File: app/ws/connectionMappings.ts [hunk 3]
```diff
@@ -25,37 +25,80 @@ const updateMappingsOnSub = (connectionId: string, channels: Set<string>, userId
 	});
 };
 
-const decreaseChannelsBinding = (connectionId: string): void => {
+const decreaseConnectionIdFromChannelBindCount = async (channel: string, connectionId: string): Promise<void> => {
+	const release = await acquireLock(channel);
+	try {
+		const listeningConnections = channelListeners.get(channel);
+		console.log('listeners ', listeningConnections, channel, connectionId);
```
Comment (LeeorShalev) [inline app/ws/connectionMappings.ts:40]: log

File: server/publications/subscription/emitter.js [hunk 4]
```diff
@@ -1,19 +1,28 @@
...
+import { msgStream } from '../../../app/lib/server/lib/msgStream';
+import { Subscriptions } from '../../../app/models';
+import { Notifications } from '../../../app/notifications';
+
+import { redisMessageHandlers } from '/app/redis/handleRedisMessage';
+import { publishToRedis } from '/app/redis/redisPublisher';
+import { settings } from '/app/settings/server';
+import ChannelHandler from '/app/ws/channelHandler';
+
+const handleSubscriptionChange = Meteor.bindEnvironment((clientAction, id, data) => {
```
Comment (LeeorShalev) [inline server/publications/subscription/emitter.js:14]: why the bind environment?

File: app/ws/channelHandler.ts [hunk 5]
```diff
@@ -23,9 +23,18 @@ const onDisconnect = (userId: string, connectionId: string): void => {
 	}, settings.get('Delay_On_Client_Disconnection') as number);
 };
 
+const addChannel = (channel: string, userId: string): void => {
+	insertChannelToMapping(channel, userId);
+};
+
```
Comment (LeeorShalev) [inline app/ws/channelHandler.ts]: it this function really necessary???
it just calls for the insertChannelToMapping

Comment (AvivBenShoham) [reply app/ws/channelHandler.ts]: You're right, I have change it to 
`const ChannelHandler = { onLogin, onDisconnect, addChannel: insertChannelToMapping, removeUserBindToRoom };`

only for readability, let me know your thoughts,

File: server/publications/subscription/emitter.js (hunk 4 above)
Comment (AvivBenShoham) [reply server/publications/subscription/emitter.js:14]: Couldn't run without it

**Pull request merged**